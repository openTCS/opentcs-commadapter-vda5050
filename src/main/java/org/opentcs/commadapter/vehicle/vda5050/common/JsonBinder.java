/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.common;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;
import java.util.function.Function;
import javax.annotation.Nonnull;

/**
 * Binds JSON strings to objects and vice versa.
 * Optionally, a filter can be applied before serializing to JSON.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  private Function<JsonNode, JsonNode> filter = Function.identity();

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Sets the filter that will be applied before serializing to JSON.
   *
   * @param filter The filter to be applied.
   */
  public void setFilter(
      @Nonnull
      Function<JsonNode, JsonNode> filter
  ) {
    this.filter = requireNonNull(filter, "filter");
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Applies the filter set via {@link #setFilter(Function)} and maps the given
   * object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalArgumentException In case there was a problem mapping the given object to JSON.
   */
  public String toJson(Object object)
      throws IllegalArgumentException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(filter.apply(objectMapper.valueToTree(object)));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalArgumentException("Could not produce JSON output", exc);
    }
  }
}
