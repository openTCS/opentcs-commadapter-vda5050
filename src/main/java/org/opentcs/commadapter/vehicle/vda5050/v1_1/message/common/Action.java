/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import java.util.List;
import javax.annotation.Nonnull;

/**
 * Action that is to be executed on a node or edge.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Action
    implements
      Serializable {

  /**
   * Name of action.
   * Identifies the function of the action.
   */
  private String actionType;
  /**
   * Unique ID to identify the action and map them to the actionState in the state.
   * Suggestion: Use UUIDs.
   */
  private String actionId;
  /**
   * [Optional] Additional information on the action.
   */
  private String actionDescription;
  /**
   * Blocking type of this action.
   */
  private BlockingType blockingType;
  /**
   * [Optional] Array of {@link ActionParameter} objects for the indicated action. (E.g. deviceId,
   * loadId or external triggers.)
   */
  private List<ActionParameter> actionParameters;

  @JsonCreator
  public Action(
      @Nonnull
      @JsonProperty(required = true, value = "actionType")
      String actionType,
      @Nonnull
      @JsonProperty(required = true, value = "actionId")
      String actionId,
      @Nonnull
      @JsonProperty(required = true, value = "blockingType")
      BlockingType blockingType
  ) {
    this.actionType = requireNonNull(actionType, "actionType");
    this.actionId = requireNonNull(actionId, "actionId");
    this.blockingType = requireNonNull(blockingType, "blockingType");
  }

  public String getActionType() {
    return actionType;
  }

  public Action setActionType(
      @Nonnull
      String actionType
  ) {
    this.actionType = requireNonNull(actionType, "actionType");
    return this;
  }

  public String getActionId() {
    return actionId;
  }

  public Action setActionId(
      @Nonnull
      String actionId
  ) {
    this.actionId = requireNonNull(actionId, "actionId");
    return this;
  }

  public String getActionDescription() {
    return actionDescription;
  }

  public Action setActionDescription(String actionDescription) {
    this.actionDescription = actionDescription;
    return this;
  }

  public BlockingType getBlockingType() {
    return blockingType;
  }

  public Action setBlockingType(BlockingType blockingType) {
    this.blockingType = blockingType;
    return this;
  }

  public List<ActionParameter> getActionParameters() {
    return actionParameters;
  }

  public Action setActionParameters(List<ActionParameter> actionParameters) {
    this.actionParameters = actionParameters;
    return this;
  }

  @Override
  public String toString() {
    return "Action{" + "actionType=" + actionType
        + ", actionId=" + actionId
        + ", actionDescription=" + actionDescription
        + ", blockingType=" + blockingType
        + ", actionParameters=" + actionParameters
        + '}';
  }

}
