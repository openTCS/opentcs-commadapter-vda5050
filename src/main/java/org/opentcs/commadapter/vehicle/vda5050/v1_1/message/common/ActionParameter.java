// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Additional parameters for an {@link Action}.
 */
public class ActionParameter
    implements
      Serializable {

  /**
   * The key of the action parameter.
   */
  private String key;
  /**
   * The value of the action parameter.
   * Can be an array, boolean, number or string.
   */
  private Object value;

  @JsonCreator
  public ActionParameter(
      @Nonnull
      @JsonProperty(required = true, value = "key")
      String key,
      @Nonnull
      @JsonProperty(required = true, value = "value")
      Object value
  ) {
    this.key = requireNonNull(key, "key");
    this.value = requireNonNull(value, "value");
  }

  public String getKey() {
    return key;
  }

  public ActionParameter setKey(
      @Nonnull
      String key
  ) {
    this.key = requireNonNull(key, "key");
    return this;
  }

  public Object getValue() {
    return value;
  }

  public ActionParameter setValue(
      @Nonnull
      Object value
  ) {
    this.value = requireNonNull(value, "value");
    return this;
  }

  @Override
  public String toString() {
    return "ActionParameter{" + "key=" + key + ", value=" + value + '}';
  }

}
