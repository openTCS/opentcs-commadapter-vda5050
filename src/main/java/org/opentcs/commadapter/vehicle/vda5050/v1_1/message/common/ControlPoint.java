// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the control points of NURBS.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ControlPoint
    implements
      Serializable {

  /**
   * X coordinate described in the world coordinate system (in m).
   */
  private Double x;
  /**
   * Y coordinate described in the world coordinate system (in m).
   */
  private Double y;
  /**
   * The weight with which this control point pulls on the curve. Range: [0 ... infinity]
   * <p>
   * When not defined, the default will be 1.0.
   */
  private Double weight;
  /**
   * [Optional] Orientation of the AGV on this position of the curve. Range: [-PI ... PI]
   * <p>
   * The orientation is in world coordinates. When not defined the orientation of the AGV will be
   * tangential to the curve.
   */
  private Double orientation;

  @JsonCreator
  public ControlPoint(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "weight")
      Double weight
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.weight
        = checkInRange(requireNonNull(weight, "weight"), 0.0, Double.MAX_VALUE, "weight");
  }

  public Double getWeight() {
    return weight;
  }

  public ControlPoint setWeight(
      @Nonnull
      Double weight
  ) {
    this.weight
        = checkInRange(requireNonNull(weight, "weight"), 0.0, Double.MAX_VALUE, "weight");
    return this;
  }

  public Double getX() {
    return x;
  }

  public ControlPoint setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public ControlPoint setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getOrientation() {
    return orientation;
  }

  public ControlPoint setOrientation(Double orientation) {
    if (orientation != null) {
      checkInRange(orientation, -Math.PI, Math.PI, "orientation");
    }
    this.orientation = orientation;
    return this;
  }

  @Override
  public String toString() {
    return "ControlPoint{" + "x=" + x
        + ", y=" + y
        + ", weight=" + weight
        + ", orientation=" + orientation
        + '}';
  }

}
