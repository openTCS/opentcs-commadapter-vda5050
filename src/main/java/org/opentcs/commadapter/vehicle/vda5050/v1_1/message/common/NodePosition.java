/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the position on a map in world coordinates.
 * <p>
 * Each floor has its own map. All maps must use the same project specific global origin.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class NodePosition
    implements
      Serializable {

  /**
   * X-position on the map in reference to the map coordinate system (in m).
   * Precision is up to the specific implementation.
   */
  private Double x;
  /**
   * Y-position on the map in reference to the map coordinate system (in m).
   * Precision is up to the specific implementation.
   */
  private Double y;
  /**
   * [Optional] Orientation of the AGV on the node (in rad). Range: [-PI ... PI]
   * <p>
   * Optional for AGVs that plan their path by themselves.
   * <p>
   * If defined, the AGV has to assume the theta angle on this node. If previous edge disallows
   * rotation, the AGV is to rotate on the node. If following edge has a differing orientation
   * defined but disallows rotation, the AGV is to rotate on the node to the edges desired rotation
   * before entering the edge.
   */
  private Double theta;
  /**
   * [Optional] Indicates how exact an AGV has to drive over a node in order for it to count as
   * traversed.
   * <p>
   * Interpretation of values:
   * <ul>
   * <li>If equals 0: No deviation is allowed. (No deviation means within the normal tolerance of
   * the AGV manufacturer.)</li>
   * <li>If greater than 0: Allowed deviation radius (in m). If the AGV passes a node within the
   * deviation radius, the node is considered to have been traversed.</li>
   * </ul>
   */
  private Double allowedDeviationXY;
  /**
   * [Optional] Indicates how big the deviation of theta angle can be. Range: [0 ... PI]
   * <p>
   * The lowest acceptable angle is {@code <theta> - <allowedDevaitionTheta>} and the highest
   * acceptable angle is {@code <theta> + <allowedDeviationTheta>}.
   */
  private Double allowedDeviationTheta;
  /**
   * Unique identification of the map in which the position is referenced.
   * <p>
   * Each map has the same origin of coordinates. When an AGV uses an elevator, e. g. leading from a
   * departure floor to a target floor, it will disappear off the map of the departure floor and
   * spawn in the related lift node on the map of the target floor.
   */
  private String mapId;
  /**
   * [Optional] Additional information on the map.
   */
  private String mapDescription;

  @JsonCreator
  public NodePosition(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "mapId")
      String mapId
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.mapId = requireNonNull(mapId, "mapId");
  }

  public Double getX() {
    return x;
  }

  public NodePosition setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public NodePosition setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getTheta() {
    return theta;
  }

  public NodePosition setTheta(Double theta) {
    if (theta != null) {
      checkInRange(theta, -Math.PI, Math.PI, "theta");
    }
    this.theta = theta;
    return this;
  }

  public Double getAllowedDeviationXY() {
    return allowedDeviationXY;
  }

  public NodePosition setAllowedDeviationXY(Double allowedDeviationXY) {
    this.allowedDeviationXY = allowedDeviationXY;
    return this;
  }

  public Double getAllowedDeviationTheta() {
    return allowedDeviationTheta;
  }

  public NodePosition setAllowedDeviationTheta(Double allowedDeviationTheta) {
    if (allowedDeviationTheta != null) {
      checkInRange(allowedDeviationTheta, 0, Math.PI, "allowedDeviationTheta");
    }
    this.allowedDeviationTheta = allowedDeviationTheta;
    return this;
  }

  public String getMapId() {
    return mapId;
  }

  public NodePosition setMapId(
      @Nonnull
      String mapId
  ) {
    this.mapId = requireNonNull(mapId, "mapId");
    return this;
  }

  public String getMapDescription() {
    return mapDescription;
  }

  public NodePosition setMapDescription(String mapDescription) {
    this.mapDescription = mapDescription;
    return this;
  }

  @Override
  public String toString() {
    return "NodePosition{" + "x=" + x
        + ", y=" + y
        + ", theta=" + theta
        + ", allowedDeviationXY=" + allowedDeviationXY
        + ", allowedDeviationTheta=" + allowedDeviationTheta
        + ", mapId=" + mapId
        + ", mapDescription=" + mapDescription
        + '}';
  }

}
