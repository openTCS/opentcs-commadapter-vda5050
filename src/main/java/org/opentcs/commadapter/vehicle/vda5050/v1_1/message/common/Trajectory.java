/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import java.util.List;
import javax.annotation.Nonnull;

/**
 * The trajectory of an AGV described as NURBS.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Trajectory
    implements
      Serializable {

  /**
   * Defines the number of control points that influence any given point on the curve.
   * Range: [1 ... infinity]
   * <p>
   * Increasing the degree increases continuity. If not defined, the default value is 1.
   */
  private Double degree;
  /**
   * Sequence of parameter values that determine where and how the control points affect the NURBS
   * curve. Range: [0.0 ... 1.0]
   * <p>
   * Has size of {@code <number of control points> + <degree> + 1}.
   */
  private List<Double> knotVector;
  /**
   * List of {@link ControlPoint} objects defining the control points of the NURBS, which include
   * the start and end point.
   */
  private List<ControlPoint> controlPoints;

  // CHECKSTYLE:OFF (Long lines because some of these parameter declarations are very long.)
  @JsonCreator
  public Trajectory(
      @Nonnull
      @JsonProperty(required = true, value = "degree")
      Double degree,
      @Nonnull
      @JsonProperty(required = true, value = "knotVector")
      List<Double> knotVector,
      @Nonnull
      @JsonProperty(required = true, value = "controlPoints")
      List<ControlPoint> controlPoints
  ) {
    this.degree
        = checkInRange(requireNonNull(degree, "degree"), 1.0, Double.MAX_VALUE, "degree");
    requireNonNull(knotVector, "knotVector");
    knotVector.forEach(value -> {
      requireNonNull(value, "knotVector value");
      checkInRange(value, 0.0, 1.0, "knotVector value");
    });
    this.knotVector = knotVector;
    this.controlPoints = requireNonNull(controlPoints, "controlPoints");
  }
  // CHECKSTYLE:ON

  public Double getDegree() {
    return degree;
  }

  public Trajectory setDegree(
      @Nonnull
      Double degree
  ) {
    this.degree
        = checkInRange(requireNonNull(degree, "degree"), 1.0, Double.MAX_VALUE, "degree");
    return this;
  }

  public List<Double> getKnotVector() {
    return knotVector;
  }

  public Trajectory setKnotVector(
      @Nonnull
      List<Double> knotVector
  ) {
    requireNonNull(knotVector, "knotVector");
    knotVector.forEach(value -> {
      requireNonNull(value, "knotVector value");
      checkInRange(value, 0.0, 1.0, "knotVector value");
    });
    this.knotVector = knotVector;
    return this;
  }

  public List<ControlPoint> getControlPoints() {
    return controlPoints;
  }

  public Trajectory setControlPoints(
      @Nonnull
      List<ControlPoint> controlPoints
  ) {
    this.controlPoints = requireNonNull(controlPoints, "controlPoints");
    return this;
  }

  @Override
  public String toString() {
    return "Trajectory{" + "degree=" + degree
        + ", knotVector=" + knotVector
        + ", controlPoints=" + controlPoints
        + '}';
  }

}
