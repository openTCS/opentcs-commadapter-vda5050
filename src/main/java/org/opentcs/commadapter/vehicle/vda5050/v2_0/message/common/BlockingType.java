// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Regulates if an {@link Action} is allowed to be executed during movement and/or parallel to other
 * actions.
 */
public enum BlockingType {

  /**
   * Action can be executed in parallel with other actions and while the vehicle is driving.
   */
  @JsonProperty(value = "NONE")
  NONE,
  /**
   * Action can be executed in parallel with other actions. Vehicle must not drive.
   */
  @JsonProperty(value = "SOFT")
  SOFT,
  /**
   * Action must not be executed in parallel with other actions. Vehicle must not drive.
   */
  @JsonProperty(value = "HARD")
  HARD;
}
