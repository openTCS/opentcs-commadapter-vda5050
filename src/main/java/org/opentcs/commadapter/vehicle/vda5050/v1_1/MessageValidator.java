/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1;

import java.io.InputStreamReader;
import java.util.Map;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonValidator;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.connection.Connection;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.instantactions.InstantActions;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.order.Order;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.state.State;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.visualization.Visualization;

/**
 * A {@link JsonValidator} for VDA5050 1.1 messages.
 */
public class MessageValidator
    extends
      JsonValidator {

  /**
   * Creates a new instance.
   *
   * @throws IllegalArgumentException If there was a problem reading the VDA5050 schemas.
   */
  public MessageValidator()
      throws IllegalArgumentException {
    super(
        Map.of(
            Connection.class,
            new InputStreamReader(
                Connection.class.getResourceAsStream(Connection.JSON_SCHEMA_PATH)
            ),
            InstantActions.class,
            new InputStreamReader(
                InstantActions.class.getResourceAsStream(InstantActions.JSON_SCHEMA_PATH)
            ),
            Order.class,
            new InputStreamReader(
                Order.class.getResourceAsStream(Order.JSON_SCHEMA_PATH)
            ),
            State.class,
            new InputStreamReader(
                State.class.getResourceAsStream(State.JSON_SCHEMA_PATH)
            ),
            Visualization.class,
            new InputStreamReader(
                Visualization.class.getResourceAsStream(Visualization.JSON_SCHEMA_PATH)
            )
        )
    );
  }

}
