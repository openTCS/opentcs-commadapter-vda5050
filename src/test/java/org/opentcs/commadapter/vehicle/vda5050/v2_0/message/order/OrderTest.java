// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.order;

import java.time.Instant;
import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.commadapter.vehicle.vda5050.ResourceLoader;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonBinder;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.MessageValidator;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.Action;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.ActionParameter;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.BlockingType;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.ControlPoint;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.NodePosition;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.Trajectory;

/**
 * Unit tests for {@link Order}.
 */
public class OrderTest {

  private static final String RESOURCE_DIR
      = "/org/opentcs/commadapter/vehicle/vda5050/v2_0/message/order/";

  private MessageValidator messageValidator;
  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    messageValidator = new MessageValidator();
    jsonBinder = new JsonBinder();
  }

  @Test
  public void validateAgainstJsonSchemaMinmal() {
    messageValidator.validate(
        jsonBinder.toJson(createOrderMinimal()),
        Order.class
    );
  }

  @Test
  public void validateAgainstJsonSchemaFull() {
    messageValidator.validate(
        jsonBinder.toJson(createOrderFull()),
        Order.class
    );
  }

  @Test
  public void validateJsonWithNullForOptionalFieldsAgainstSchema() {
    messageValidator.validate(
        ResourceLoader.load(RESOURCE_DIR + "orderMessageWithNullForOptionalFields.json"),
        Order.class
    );
  }

  @Test
  public void jsonSampleMinimal() {
    Approvals.verify(jsonBinder.toJson(createOrderMinimal()));
  }

  @Test
  public void jsonSampleFull() {
    Approvals.verify(jsonBinder.toJson(createOrderFull()));
  }

  private static Order createOrderMinimal() {
    return new Order(
        0L,
        Instant.EPOCH,
        "version",
        "manufacture",
        "serial-number",
        "some-order-id",
        0L,
        List.of(
            new Node(
                "source-node",
                0L,
                true,
                List.of()
            )
        ),
        List.of()
    );
  }

  private static Order createOrderFull() {
    return new Order(
        0L,
        Instant.EPOCH,
        "version",
        "manufacture",
        "serial-number",
        "some-order-id",
        0L,
        List.of(
            new Node(
                "source-node",
                0L,
                true,
                List.of(
                    new Action("some-action-type", "some-action-id", BlockingType.NONE)
                        .setActionDescription("some-action-description")
                        .setActionParameters(
                            List.of(
                                new ActionParameter("some-key", "some-value"),
                                new ActionParameter("some-other-key", "some-other-value")
                            )
                        )
                )
            )
                .setNodeDescription("some-node-description")
                .setNodePosition(
                    new NodePosition(1.2, 3.4, "some-map-id")
                        .setTheta(1.3)
                        .setMapDescription("some-map-description")
                        .setAllowedDeviationXY(0.1)
                        .setAllowedDeviationTheta(0.1)
                ),
            new Node(
                "destination-node",
                0L,
                true,
                List.of(
                    new Action("some-action-type", "some-action-id", BlockingType.NONE)
                        .setActionDescription("some-action-description")
                        .setActionParameters(
                            List.of(
                                new ActionParameter("some-key", "some-value"),
                                new ActionParameter("some-other-key", "some-other-value")
                            )
                        )
                )
            )
                .setNodeDescription("some-node-description")
                .setNodePosition(
                    new NodePosition(5.6, 7.8, "some-map-id")
                        .setTheta(0.0)
                        .setMapDescription("some-map-description")
                        .setAllowedDeviationXY(0.2)
                        .setAllowedDeviationTheta(0.3)
                )
        ),
        List.of(
            new Edge(
                "some-edge-id",
                0L,
                true,
                "source-node",
                "destination-node",
                List.of(
                    new Action("some-action-type", "some-action-id", BlockingType.NONE)
                        .setActionDescription("some-action-description")
                        .setActionParameters(
                            List.of(
                                new ActionParameter("some-key", "some-value"),
                                new ActionParameter("some-other-key", "some-other-value")
                            )
                        )
                )
            )
                .setDirection("some-direction")
                .setLength(123.0)
                .setEdgeDescription("some-edge-description")
                .setMaxHeight(3.0)
                .setMaxRotationSpeed(44.0)
                .setMaxSpeed(123.0)
                .setMinHeight(1.0)
                .setOrientation(0.0)
                .setRotationAllowed(true)
                .setTrajectory(
                    new Trajectory(
                        1.0,
                        List.of(0.1, 0.2, 0.3),
                        List.of(
                            new ControlPoint(1.2, 2.3)
                                .setWeight(1.0)
                        )
                    )
                )
        )
    )
        .setZoneSetId("some-zone-set-id");
  }

}
