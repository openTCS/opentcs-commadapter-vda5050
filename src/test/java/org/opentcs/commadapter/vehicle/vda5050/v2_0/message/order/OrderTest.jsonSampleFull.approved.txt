{
  "headerId" : 0,
  "timestamp" : "1970-01-01T00:00:00Z",
  "version" : "version",
  "manufacturer" : "manufacture",
  "serialNumber" : "serial-number",
  "orderId" : "some-order-id",
  "orderUpdateId" : 0,
  "nodes" : [ {
    "nodeId" : "source-node",
    "sequenceId" : 0,
    "released" : true,
    "actions" : [ {
      "actionType" : "some-action-type",
      "actionId" : "some-action-id",
      "blockingType" : "NONE",
      "actionDescription" : "some-action-description",
      "actionParameters" : [ {
        "key" : "some-key",
        "value" : "some-value"
      }, {
        "key" : "some-other-key",
        "value" : "some-other-value"
      } ]
    } ],
    "nodeDescription" : "some-node-description",
    "nodePosition" : {
      "x" : 1.2,
      "y" : 3.4,
      "mapId" : "some-map-id",
      "theta" : 1.3,
      "allowedDeviationXY" : 0.1,
      "allowedDeviationTheta" : 0.1,
      "mapDescription" : "some-map-description"
    }
  }, {
    "nodeId" : "destination-node",
    "sequenceId" : 0,
    "released" : true,
    "actions" : [ {
      "actionType" : "some-action-type",
      "actionId" : "some-action-id",
      "blockingType" : "NONE",
      "actionDescription" : "some-action-description",
      "actionParameters" : [ {
        "key" : "some-key",
        "value" : "some-value"
      }, {
        "key" : "some-other-key",
        "value" : "some-other-value"
      } ]
    } ],
    "nodeDescription" : "some-node-description",
    "nodePosition" : {
      "x" : 5.6,
      "y" : 7.8,
      "mapId" : "some-map-id",
      "theta" : 0.0,
      "allowedDeviationXY" : 0.2,
      "allowedDeviationTheta" : 0.3,
      "mapDescription" : "some-map-description"
    }
  } ],
  "edges" : [ {
    "edgeId" : "some-edge-id",
    "sequenceId" : 0,
    "released" : true,
    "startNodeId" : "source-node",
    "endNodeId" : "destination-node",
    "actions" : [ {
      "actionType" : "some-action-type",
      "actionId" : "some-action-id",
      "blockingType" : "NONE",
      "actionDescription" : "some-action-description",
      "actionParameters" : [ {
        "key" : "some-key",
        "value" : "some-value"
      }, {
        "key" : "some-other-key",
        "value" : "some-other-value"
      } ]
    } ],
    "edgeDescription" : "some-edge-description",
    "maxSpeed" : 123.0,
    "maxHeight" : 3.0,
    "minHeight" : 1.0,
    "orientation" : 0.0,
    "direction" : "some-direction",
    "rotationAllowed" : true,
    "maxRotationSpeed" : 44.0,
    "length" : 123.0,
    "trajectory" : {
      "degree" : 1.0,
      "knotVector" : [ 0.1, 0.2, 0.3 ],
      "controlPoints" : [ {
        "x" : 1.2,
        "y" : 2.3,
        "weight" : 1.0
      } ]
    }
  } ],
  "zoneSetId" : "some-zone-set-id"
}